garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,206 @@
+                    self._optimize_loss = optimizer.minimize(
+                        self._loss, var_list=self.qf.get_trainable_vars())
+
+            self._train_qf = tensor_utils.compile_function(
+                inputs=[
+                    self.qf.input, self.action_t_ph, self.reward_t_ph,
+                    self.done_t_ph, self.target_qf.input
+                ],
+                outputs=[self._loss, self._optimize_loss])
+
+    def train_once(self, itr, paths):
+        """Train the algorithm once."""
+        epoch = itr / self.n_epoch_cycles
+
+        self.episode_rewards.extend(paths['undiscounted_returns'])
+        last_average_return = np.mean(self.episode_rewards)
+        for train_itr in range(self.n_train_steps):
+            if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+                self.evaluate = True
+                qf_loss = self.optimize_policy(epoch, None)
+                self.episode_qf_losses.append(qf_loss)
+
+        if self.evaluate:
+            if itr % self.target_network_update_freq == 0:
+                self._qf_update_ops()
+
+        if itr % self.n_epoch_cycles == 0:
+            if self.evaluate:
+                mean100ep_rewards = round(
+                    np.mean(self.episode_rewards[-100:]), 1)
+                mean100ep_qf_loss = np.mean(self.episode_qf_losses[-100:])
+                tabular.record('Epoch', epoch)
+                tabular.record('AverageReturn', np.mean(self.episode_rewards))
+                tabular.record('StdReturn', np.std(self.episode_rewards))
+                tabular.record('Episode100RewardMean', mean100ep_rewards)
+                tabular.record('{}/Episode100LossMean'.format(self.qf.name),
+                               mean100ep_qf_loss)
+        return last_average_return
+
+    @overrides
+    def get_itr_snapshot(self, itr):
+        """Get snapshot of the policy."""
+        return dict(itr=itr, policy=self.policy)
+
+    @overrides
+    def optimize_policy(self, itr, sample_data):
+        """Optimize network using experiences from replay buffer."""
+        transitions = self.replay_buffer.sample(self.buffer_batch_size)
+
+        observations = transitions['observation']
+        rewards = transitions['reward']
+        actions = transitions['action']
+        next_observations = transitions['next_observation']
+        dones = transitions['terminal']
+
+        # normalize pixel to range [0, 1] since the samples stored in the
+        # replay buffer are of type uint8 and not normalized, for memory
+        # optimization
+        observations = normalize_pixel_batch(self.env_spec, observations)
+        next_observations = normalize_pixel_batch(self.env_spec,
+                                                  next_observations)
+        loss, _ = self._train_qf(observations, actions, rewards, dones,
+                                 next_observations)
+
+        return loss
+
+
+def huber_loss(x, delta=1.0):
```

Comment:
this exists in tensorflow: https://www.tensorflow.org/api_docs/python/tf/losses/huber_loss

is their version not useful?

Reply:
I will use it.